cogito/put.go

```diff
@@ -26,91 +45,167 @@ import (
 // Additionally, the script may emit metadata as a list of key-value pairs. This data is
 // intended for public consumption and will make it upstream, intended to be shown on the
 // build's page.
-func Put(log hclog.Logger, in io.Reader, out io.Writer, args []string) error {
-	var pi PutInput
+func Put(log hclog.Logger, in io.Reader, out io.Writer, args []string, putter Putter) error {
+	if err := putter.LoadConfiguration(in, args); err != nil {
+		return fmt.Errorf("put: %s", err)
+	}
+
+	if err := putter.ProcessInputDir(); err != nil {
+		return fmt.Errorf("put: %s", err)
+	}
+
+	// We invoke all the sinks and keep going also if some of them return an error.
+	var sinkErrors []error
+	for _, sink := range putter.Sinks() {
+		if err := sink.Send(); err != nil {
+			sinkErrors = append(sinkErrors, err)
+		}
+	}
+	if len(sinkErrors) > 0 {
+		return fmt.Errorf("put: %s", multiErrString(sinkErrors))
+	}
+
+	if err := putter.Output(out); err != nil {
+		return fmt.Errorf("put: %s", err)
+	}
+
+	return nil
+}
+
+// ProdPutter is an implementation of a [Putter] for the Cogito resource.
+// Use [NewPutter] to create an instance.
+type ProdPutter struct {
+	ghAPI  string
+	log    hclog.Logger
+	gitRef string
+
+	Pi       PutInput
+	InputDir string
+}
+
+// NewPutter returns a Cogito ProdPutter.
+func NewPutter(ghAPI string, log hclog.Logger) *ProdPutter {
+	return &ProdPutter{
+		ghAPI: ghAPI,
+		log:   log,
+	}
+}
+
+func (pu *ProdPutter) LoadConfiguration(in io.Reader, args []string) error {
 	dec := json.NewDecoder(in)
 	dec.DisallowUnknownFields()
-	if err := dec.Decode(&pi); err != nil {
+	if err := dec.Decode(&pu.Pi); err != nil {
```

Comment:
At a certain point you will have to accept what is idiomatic Go and not try to write Python in Go or to think Python while reading Go code :-)

The same as I accept the in my opinion error-prone and added cognitive load of Python functions when one never knows wether they raise an exception or return an error, and the only way to know is to read the whole called function in a recursive never ending way. This is why Go does error handling the way it does it: impossible to forget to handle an exception, since they do not exist. The only thing you need to look at in order to correctly handle a function call is to read the function type :-)

Going into the specifics of the code here:

`Pi` cannot be called a `Request` because it is more than a Request, it also contains the environment variables. This is why I called it `PutInput`. In this case, in my opinion the error is on the Concourse protocol design, since they pass "input" in 3 channels:
1. stdin (the JSON object)
2. command-line arguments (the directory which contains the resource inputs volumes)
3. environment variables (build number, pipeline name etc)
while I think it would have made more sense to encode everything in a JSON object...

We could not even tell ourselves that we consider that these 3 "inputs" are still part of one request, because this request is actually different in the 3 steps, so we would have to call it `PutRequest`, `GetRequest` and `CheckRequest`. Which is what we have now with `PutInput`, `GetInput` and `CheckInput`. I am not satisfied either with the `Input` suffix, since it can be confused with the "input directory".

If you think that here replacing the `Input` suffix with `Request` would make it less-confusing, I am OK with renaming. Once we decide, we could have one the following

- pu.PutInput
- pu.PutRequest

Reply:
Yes, using `PutRequest`, `GetRequest` and `CheckRequest` would be much better!

Reply:
Types: switched to `PutRequest`, `GetRequest` and `CheckRequest`.
Variables: from `ci`, `Pi`, `gi` to `request` (context is clear enough to understand which request it is).

It is now more readable indeed. Thanks!